Add support for signing in with a recovery code

monfresh opened the pull request:
**Why**: To allow users who lose access to their 2FA phone to recover
access to their account.

**How**:
- Follow the design pattern of keeping each method of authentication in
its own controller, and a form object to validate the code.
- Generate a 16-character alphanumeric random code and one-way encrypt
it in the database. This has the advantage of keeping the code secure,
but the disadvantage that it cannot be displayed to the user more than
once, which could be problematic if they generate multiple codes and
don't remember which one is the latest one.
- When a user first signs up, after they confirm their 2FA phone, they
will be presented with their recovery code with instructions for how
and where to keep it safe.
- Once signed in, a user can generate a new code. Note that this
currently doesn't include any messages about what that implies.
- When a user signs in with their recovery code, a new one is
immediately generated, which replaces their previous one.

monfresh commented:
Please merge #473 first before reviewing this, as this was based on that branch. Once 473 is merged, I will rebase, and this PR will become smaller than it is now.

monfresh commented:
Rebased and ready for review now. Thanks!

monfresh commented:
Screenshots:

Link to recovery code fallback:
<img width="809" alt="screen shot 2016-09-16 at 4 27 51 pm" src="https://cloud.githubusercontent.com/assets/811150/18643524/8af8fb9c-7e73-11e6-97ba-735d17730511.png">

After signing in with 2FA for the first time, or after signing in via recovery code:
<img width="628" alt="screen shot 2016-09-19 at 2 31 20 pm" src="https://cloud.githubusercontent.com/assets/811150/18644041/c768fe72-7e75-11e6-8e25-0fe18b6af7db.png">

Button to regenerate the recovery code from the profile page:
<img width="683" alt="screen shot 2016-09-19 at 2 32 20 pm" src="https://cloud.githubusercontent.com/assets/811150/18644066/e52f582a-7e75-11e6-9c22-2a6097a5fe0f.png">

monfresh commented:
Found a bug. Fix forthcoming.

monfresh commented:
Note that this still needs feedback from the design team, which we will incorporate in subsequent PRs.

pkarman commented on app/controllers/concerns/two_factor_authenticatable.rb:
```diff
@@ -46,12 +46,12 @@ def handle_valid_otp
   end
 
-  def handle_invalid_otp
+  def handle_invalid_otp(type: 'otp')
```
since the method name has `_otp` in it, maybe rename method, or add a comment about what other valid `type` values are supported?

monfresh replied:
I'll add a comment. I didn't want to rename the method to keep the changes small, but I agree this should be renamed in a future PR. There is more refactoring to done around 2FA confirmation.

pkarman commented on app/services/recovery_code_generator.rb:
```diff
@@ -0,0 +1,28 @@
+  end
+
+  def raw_recovery_code
+    @raw_recovery_code ||= SecureRandom.hex(recovery_code_length / 2)
```
must `recovery_code_length` be an even number?

monfresh replied:
It doesn't have to be, but it could lead to confusion if someone sets it to an odd number, but the code length is even. I'll add a comment to application.yml

pkarman commented on config/application.yml.example:
```diff
@@ -12,6 +12,7 @@ email_from: 'no-reply@login.gov'
 idv_max_attempts: '3'
 idv_attempt_window_in_hours: '24'
 mailer_domain_name: 'http://localhost:3000'
+recovery_code_length: '16'
```
maybe comment about min/max values? even vs odd?

monfresh commented:
To test:

**Scenario 1**:
- Sign up with a new account
- [ ] After confirming 2FA phone, expect to see a recovery code screen that displays your recovery code and instructions for keeping it in a safe place. Copy the code to your clipboard so you can paste it in the Scenario 2 below.
- Click "Continue"
- [ ] Expect that the next screen is the profile page. Make sure the screen did not refresh to show a new recovery code.

**Scenario 2**:
- Sign out of the above account, and sign back in with just email and password
- On the screen where you can choose how you want to receive your OTP code, click the "recovery code" link at the bottom
- Enter your recovery code and submit the form
- [ ] Expect that you are successfully signed in and that you see a recovery code screen that displays your new recovery code and instructions for keeping it in a safe place.
- Click "Continue"
- [ ] Expect that the next screen is the profile page

**Scenario 3**:
- Sign in with an account that has a recovery code and go to your profile page
- Click on the `Regenerate` button
- [ ] Expect to see a recovery code screen that displays your new recovery code and instructions for keeping it in a safe place.
- Click "Continue"
- [ ] Expect that the next screen is the profile page

pkarman commented:
@monfresh the scenarios you outline for testing look good: can they be turned into feature specs?

monfresh commented:
The bug I found only happens with the test users that we prepopulate via the seeds file. The bug is that after you sign in with one of the test users, you see the recovery code screen, but clicking continue displays the screen again with a new code. This doesn't seem to affect new users, but I think it will affect the users we migrate from USCIS. Looking into it now.

monfresh commented:
One more scenario:

Make sure that entering an invalid recovery code doesn't sign you in and that it locks you out after 3 failed attempts, just like for OTP.

monfresh commented:
I addressed the PR comments and fixed the bug. Check out my second commit please. Note that the bug fix requires code duplication until I refactor the phone confirmation code.

pkarman commented:
feature tests lgtm -- thanks @monfresh
